### PR TITLE
fix bug with how we unregistered host memory

### PR DESCRIPTION
**Pull request by stonea**
The gpu runtime keeps a list of host registered variables (variables stored in the CPU's ram that we want to directly access by the gpu), which we unregister after a kernel launch is complete. This list consists of double pointers and I was accidentally unregistering the outer rather than the inner pointer.

This causes the following error:

`internal error: gpu-nvidia.c:388: Error calling CUDA function: pointer does not correspond to a registered memory region (Code: 713)`

*Edit*: But even more than that, for AMD this list contains pointers that are only meant to be used by the device. When we call the unregister function we need to pass in the host "version" of that pointer so this PR changes things so we keep a list of the host version of these pointers.

TODO:

- [x] paratest NVIDIA
- [x] paratest AMD

### CODE REVIEW

**Review by DanilaFe — APPROVED** [review]

File: runtime/src/chpl-gpu.c [hunk 1]
```diff
@@ -637,7 +637,7 @@ void chpl_gpu_deinit_kernel_cfg(void* _cfg) {
   chpl_mem_free(cfg->reduce_vars, cfg->ln, cfg->fn);
 
   for (int i=0 ; i<cfg->n_host_registered; i++) {
-    chpl_gpu_impl_host_unregister(cfg->host_registered_vars[i]);
+    chpl_gpu_impl_host_unregister(*(cfg->host_registered_vars[i]));
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:640]: My favorite kind of patch.

**Review by e-kayrakli — APPROVED** [review]

File: runtime/src/chpl-gpu.c [hunk 2]
```diff
@@ -712,6 +718,7 @@ void chpl_gpu_arg_host_register(void* _cfg, void* arg, size_t size) {
   void *dev_arg = chpl_gpu_impl_host_register(*((void**)arg), size);
   *(cfg->host_registered_vars[cfg->cur_host_registered_var]) = dev_arg;
   cfg_add_direct_param(cfg, cfg->host_registered_vars[cfg->cur_host_registered_var]);
+  cfg->host_registered_vars_host_ptrs[cfg->cur_host_registered_var] = *((void**)arg);
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:721]: gah... so unfortunate that these APIs don't have an interface for getting a host pointer from a device one for this scenario...

File: runtime/src/chpl-gpu.c [hunk 3]
```diff
@@ -257,8 +257,10 @@ typedef struct kernel_cfg_s {
   void*** kernel_params;
   int n_host_registered;
   int cur_host_registered_var;
-  void** host_registered_var_boxes;
-  void*** host_registered_vars;
+  void** host_registered_var_boxes; // array of device pointers
+  void*** host_registered_vars;     // array of pointers to device pointer
+                                    // (stored in host_registered_var_boxes)
+  void** host_registered_vars_host_ptrs; // array of host pointers
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:263]: I am slowly getting a negative feeling for the long variable names here :D I'm OK for now, but I may not be for the next PR :)

**Review by e-kayrakli — APPROVED** [review]

File: runtime/src/chpl-gpu.c (hunk 3 above)
Comment (stonea) [reply runtime/src/chpl-gpu.c:263]: I_blame_snake_case.

**Pull request merged**